File: substrate/collective_multiselect_floating_dr_draft_before_merge_compose.py
"""Collective multiselect over floating DR draft-before-merge MO price-ceiling (pure).

live_dispatched / pack_dispatched / merge_executed / analysis_written always False.
twin_written / draft_written / live_execution_authorized always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.floating_dr_draft_before_merge_mo_price_ceiling_compose import (
    FloatingDrDraftBeforeMergeMoPriceCeilingCompose,
    FloatingDrDraftBeforeMergeMoPriceCeilingComposeError,
    compose_floating_dr_draft_before_merge_mo_price_ceiling,
)
from substrate.floating_multi_select_collective_cohesive_compose import (
    FloatingMultiSelectCollectiveCohesiveCompose,
    FloatingMultiSelectCollectiveCohesiveComposeError,
    compose_floating_multi_select_collective_cohesive,
)


class CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(ValueError):
    """Fail-closed validation for collective multiselect + floating DR pack."""


@dataclass(frozen=True)
class CollectiveMultiselectFloatingDrDraftBeforeMergeCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    asset_id: str
    title: str
    account_id: str
    multiselect: FloatingMultiSelectCollectiveCohesiveCompose
    floating_dr_pack: FloatingDrDraftBeforeMergeMoPriceCeilingCompose
    session_aligned: bool
    parent_aligned: bool
    pack_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    analysis_written: bool
    twin_written: bool
    draft_written: bool
    live_execution_authorized: bool
    charge_executed: bool
    prompts_injected: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    remote_index_queried: bool
    backlog_mutated: bool
    store_mutated: bool
    suite_rewritten: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    live_dispatch_authorized: bool
    inventory_mutated: bool
    record_persisted: bool
    purchase_executed: bool
    hosted: bool
    remote_fetched: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "multiselect": self.multiselect.to_dict(),
            "floating_dr_pack": self.floating_dr_pack.to_dict(),
            "session_aligned": self.session_aligned,
            "parent_aligned": self.parent_aligned,
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "analysis_written": False,
            "twin_written": False,
            "draft_written": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "remote_index_queried": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "live_dispatch_authorized": False,
            "inventory_mutated": False,
            "record_persisted": False,
            "purchase_executed": False,
            "hosted": False,
            "remote_fetched": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "collective_multiselect_floating_dr_draft_before_merge_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_collective_multiselect_floating_dr_draft_before_merge(
    *,
    multiselect: object,
    floating_dr_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> CollectiveMultiselectFloatingDrDraftBeforeMergeCompose:
    """Collective multiselect on floating DR draft-before-merge. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(multiselect, dict):
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            "multiselect must be an object"
        )
    if not isinstance(floating_dr_pack, dict):
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            "floating_dr_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · pack_dispatched=false · merge_executed=false",
        "analysis_written=false · twin_written=false · draft_written=false",
        "production_router_verdict=REJECT",
    ]

    try:
        ms = compose_floating_multi_select_collective_cohesive(
            session_id=multiselect.get("session_id"),
            parent_asset_id=multiselect.get("parent_asset_id"),
            members=multiselect.get("members"),
            selected_instance_ids=multiselect.get("selected_instance_ids"),
            pack_mode=multiselect.get("pack_mode"),
            cohesive_prompt=multiselect.get("cohesive_prompt"),
            operator_ack=operator_ack,
            extra_context=multiselect.get("extra_context"),
            analysis_kind=multiselect.get("analysis_kind"),
            extra_findings=multiselect.get("extra_findings"),
        )
    except FloatingMultiSelectCollectiveCohesiveComposeError as e:
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            str(e)
        ) from e
    notes.extend(f"[multiselect] {n}" for n in ms.notes)

    try:
        fdr = compose_floating_dr_draft_before_merge_mo_price_ceiling(
            highlight_surface=floating_dr_pack.get("highlight_surface"),
            draft_pack=floating_dr_pack.get("draft_pack"),
            operator_ack=operator_ack,
            require_both=floating_dr_pack.get("require_both"),
        )
    except FloatingDrDraftBeforeMergeMoPriceCeilingComposeError as e:
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            str(e)
        ) from e
    notes.extend(f"[floating_dr_pack] {n}" for n in fdr.notes)

    session = _require_nonempty(ms.session_id, field="session_id")
    parent = _require_nonempty(ms.parent_asset_id, field="parent_asset_id")
    week = _require_nonempty(fdr.week_id, field="week_id")
    asset = _require_nonempty(fdr.asset_id, field="asset_id")
    title = _require_nonempty(fdr.title, field="title")
    account = _require_nonempty(fdr.account_id, field="account_id")

    session_aligned = fdr.session_id == session
    parent_aligned = fdr.parent_asset_id == parent or fdr.asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between multiselect and floating_dr_pack — "
            "pack_ready blocked"
        )
    else:
        notes.append("session_aligned=true")
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between multiselect and floating_dr_pack — "
            "pack_ready blocked"
        )
    else:
        notes.append("parent_aligned=true")

    if require:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and ms.pack_ready is True
            and fdr.pack_ready is True
            and fdr.production_router_verdict == "REJECT"
            and ms.live_dispatched is False
            and ms.pack_dispatched is False
            and ms.merge_executed is False
            and ms.analysis_written is False
            and fdr.live_dispatched is False
            and fdr.merge_executed is False
            and fdr.twin_written is False
            and fdr.draft_written is False
            and fdr.live_execution_authorized is False
            and fdr.charge_executed is False
            and fdr.remote_index_queried is False
            and fdr.pdf_primary is False
            and fdr.live_router_authorized is False
            and fdr.secrets_stored is False
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and operator_ack is True
            and fdr.production_router_verdict == "REJECT"
            and fdr.pdf_primary is False
            and ms.live_dispatched is False
            and (ms.pack_ready is True or fdr.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — collective multiselect + floating DR "
            "draft-before-merge ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — multiselect, floating_dr_pack, alignment, or "
            "operator_ack gate open"
        )

    if (
        ms.live_dispatched is not False
        or ms.pack_dispatched is not False
        or ms.merge_executed is not False
        or ms.analysis_written is not False
        or fdr.live_dispatched is not False
        or fdr.merge_executed is not False
        or fdr.twin_written is not False
        or fdr.draft_written is not False
        or fdr.live_execution_authorized is not False
        or fdr.charge_executed is not False
        or fdr.remote_index_queried is not False
        or fdr.pdf_primary is not False
        or fdr.live_router_authorized is not False
        or fdr.secrets_stored is not False
        or fdr.production_router_verdict != "REJECT"
    ):
        raise CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "analysis_written=false",
            "twin_written=false",
            "draft_written=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "prompts_injected=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "remote_index_queried=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "live_dispatch_authorized=false",
            "inventory_mutated=false",
            "record_persisted=false",
            "purchase_executed=false",
            "hosted=false",
            "remote_fetched=false",
            "production_router_verdict=REJECT",
        )
    )

    return CollectiveMultiselectFloatingDrDraftBeforeMergeCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        asset_id=asset,
        title=title,
        account_id=account,
        multiselect=ms,
        floating_dr_pack=fdr,
        session_aligned=session_aligned,
        parent_aligned=parent_aligned,
        pack_ready=pack_ready,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        analysis_written=False,
        twin_written=False,
        draft_written=False,
        live_execution_authorized=False,
        charge_executed=False,
        prompts_injected=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        remote_index_queried=False,
        backlog_mutated=False,
        store_mutated=False,
        suite_rewritten=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        live_dispatch_authorized=False,
        inventory_mutated=False,
        record_persisted=False,
        purchase_executed=False,
        hosted=False,
        remote_fetched=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "collective_multiselect_floating_dr_draft_before_merge_compose_advisory"
        ),
    )


def format_collective_multiselect_floating_dr_draft_before_merge_summary(
    c: CollectiveMultiselectFloatingDrDraftBeforeMergeCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"multi_ready={c.multiselect.pack_ready} · "
        f"float_ready={c.floating_dr_pack.pack_ready} · "
        f"mode={c.multiselect.pack_mode} · "
        f"session_aligned={c.session_aligned} · "
        f"parent_aligned={c.parent_aligned} · "
        f"verdict={c.production_router_verdict} · "
        "live_dispatched=false · pack_dispatched=false · analysis_written=false"
    )


__all__ = [
    "CollectiveMultiselectFloatingDrDraftBeforeMergeCompose",
    "CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError",
    "compose_collective_multiselect_floating_dr_draft_before_merge",
    "format_collective_multiselect_floating_dr_draft_before_merge_summary",
]

File: apps/reading/src/api/collectiveMultiselectFloatingDrDraftBeforeMergeCompose.ts
/**
 * Floating multi-select collective cohesive residual over floating DR +
 * draft-before-merge + MO price-ceiling pack (pure).
 *
 * Operator vision: click multiple floating/sub-agent deep research instances
 * and engage them as one cohesive unit while floating DR from highlights,
 * draft-before-merge, and Midnight Oil price-ceiling honesty remain pure —
 * without live dispatch, pack dispatch, analysis writes, or parent merges.
 *
 * live_dispatched / pack_dispatched / merge_executed / analysis_written always false.
 * twin_written / draft_written / live_execution_authorized always false.
 * production_router_verdict always REJECT.
 */

import {
  composeFloatingMultiSelectCollectiveCohesive,
  type FloatingMultiSelectCollectiveCohesiveCompose,
  type FloatingMultiSelectCollectiveCohesiveInput,
} from "./floatingMultiSelectCollectiveCohesiveCompose";
import {
  composeFloatingDrDraftBeforeMergeMoPriceCeiling,
  type FloatingDrDraftBeforeMergeMoPriceCeilingCompose,
  type FloatingDrDraftBeforeMergeMoPriceCeilingInput,
} from "./floatingDrDraftBeforeMergeMoPriceCeilingCompose";

export interface CollectiveMultiselectFloatingDrDraftBeforeMergeInput {
  multiselect: Omit<FloatingMultiSelectCollectiveCohesiveInput, "operator_ack">;
  floating_dr_pack: Omit<
    FloatingDrDraftBeforeMergeMoPriceCeilingInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface CollectiveMultiselectFloatingDrDraftBeforeMergeCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  multiselect: FloatingMultiSelectCollectiveCohesiveCompose;
  floating_dr_pack: FloatingDrDraftBeforeMergeMoPriceCeilingCompose;
  session_aligned: boolean;
  parent_aligned: boolean;
  pack_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  analysis_written: false;
  twin_written: false;
  draft_written: false;
  live_execution_authorized: false;
  charge_executed: false;
  prompts_injected: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  remote_index_queried: false;
  backlog_mutated: false;
  store_mutated: false;
  suite_rewritten: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  live_dispatch_authorized: false;
  inventory_mutated: false;
  record_persisted: false;
  purchase_executed: false;
  hosted: false;
  remote_fetched: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "collective_multiselect_floating_dr_draft_before_merge_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Multi-select collective cohesive stacked on floating DR draft-before-merge MO pack.
 * Never dispatches packs; never writes analysis; ND REJECT.
 */
export function composeCollectiveMultiselectFloatingDrDraftBeforeMerge(
  input: CollectiveMultiselectFloatingDrDraftBeforeMergeInput,
): CollectiveMultiselectFloatingDrDraftBeforeMergeCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.multiselect || typeof input.multiselect !== "object") {
    throw new Error("multiselect must be an object");
  }
  if (!input.floating_dr_pack || typeof input.floating_dr_pack !== "object") {
    throw new Error("floating_dr_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · pack_dispatched=false · merge_executed=false",
    "analysis_written=false · twin_written=false · draft_written=false",
    "production_router_verdict=REJECT",
  ];

  const multiselect = composeFloatingMultiSelectCollectiveCohesive({
    ...input.multiselect,
    operator_ack: input.operator_ack,
  });
  notes.push(...multiselect.notes.map((n) => `[multiselect] ${n}`));

  const floating_dr_pack = composeFloatingDrDraftBeforeMergeMoPriceCeiling({
    ...input.floating_dr_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(...floating_dr_pack.notes.map((n) => `[floating_dr_pack] ${n}`));

  const session_id = requireNonEmpty(multiselect.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    multiselect.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(floating_dr_pack.week_id, "week_id");
  const asset_id = requireNonEmpty(floating_dr_pack.asset_id, "asset_id");
  const title = requireNonEmpty(floating_dr_pack.title, "title");
  const account_id = requireNonEmpty(floating_dr_pack.account_id, "account_id");

  const session_aligned = floating_dr_pack.session_id === session_id;
  const parent_aligned =
    floating_dr_pack.parent_asset_id === parent_asset_id ||
    floating_dr_pack.asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between multiselect and floating_dr_pack — pack_ready blocked",
    );
  } else {
    notes.push("session_aligned=true");
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between multiselect and floating_dr_pack — pack_ready blocked",
    );
  } else {
    notes.push("parent_aligned=true");
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      multiselect.pack_ready === true &&
      floating_dr_pack.pack_ready === true &&
      floating_dr_pack.production_router_verdict === "REJECT" &&
      multiselect.live_dispatched === false &&
      multiselect.pack_dispatched === false &&
      multiselect.merge_executed === false &&
      multiselect.analysis_written === false &&
      floating_dr_pack.live_dispatched === false &&
      floating_dr_pack.merge_executed === false &&
      floating_dr_pack.twin_written === false &&
      floating_dr_pack.draft_written === false &&
      floating_dr_pack.live_execution_authorized === false &&
      floating_dr_pack.charge_executed === false &&
      floating_dr_pack.remote_index_queried === false &&
      floating_dr_pack.pdf_primary === false &&
      floating_dr_pack.live_router_authorized === false &&
      floating_dr_pack.secrets_stored === false &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      floating_dr_pack.production_router_verdict === "REJECT" &&
      floating_dr_pack.pdf_primary === false &&
      multiselect.live_dispatched === false &&
      (multiselect.pack_ready === true || floating_dr_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — collective multiselect + floating DR draft-before-merge ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — multiselect, floating_dr_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    multiselect.live_dispatched !== false ||
    multiselect.pack_dispatched !== false ||
    multiselect.merge_executed !== false ||
    multiselect.analysis_written !== false ||
    floating_dr_pack.live_dispatched !== false ||
    floating_dr_pack.merge_executed !== false ||
    floating_dr_pack.twin_written !== false ||
    floating_dr_pack.draft_written !== false ||
    floating_dr_pack.live_execution_authorized !== false ||
    floating_dr_pack.charge_executed !== false ||
    floating_dr_pack.remote_index_queried !== false ||
    floating_dr_pack.pdf_primary !== false ||
    floating_dr_pack.live_router_authorized !== false ||
    floating_dr_pack.secrets_stored !== false ||
    floating_dr_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("analysis_written=false");
  notes.push("twin_written=false");
  notes.push("draft_written=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("remote_index_queried=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("inventory_mutated=false");
  notes.push("record_persisted=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("remote_fetched=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    week_id,
    asset_id,
    title,
    account_id,
    multiselect,
    floating_dr_pack,
    session_aligned,
    parent_aligned,
    pack_ready,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    analysis_written: false,
    twin_written: false,
    draft_written: false,
    live_execution_authorized: false,
    charge_executed: false,
    prompts_injected: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    remote_index_queried: false,
    backlog_mutated: false,
    store_mutated: false,
    suite_rewritten: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    live_dispatch_authorized: false,
    inventory_mutated: false,
    record_persisted: false,
    purchase_executed: false,
    hosted: false,
    remote_fetched: false,
    production_router_verdict: "REJECT",
    notes,
    authority:
      "collective_multiselect_floating_dr_draft_before_merge_compose_advisory",
  };
}

export function formatCollectiveMultiselectFloatingDrDraftBeforeMergeSummary(
  c: CollectiveMultiselectFloatingDrDraftBeforeMergeCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `multi_ready=${c.multiselect.pack_ready} · ` +
    `float_ready=${c.floating_dr_pack.pack_ready} · ` +
    `mode=${c.multiselect.pack_mode} · ` +
    `session_aligned=${c.session_aligned} · ` +
    `parent_aligned=${c.parent_aligned} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_dispatched=false · pack_dispatched=false · analysis_written=false`
  );
}
